xfreerdp: Add -X option to allow embedding.

From: Thierry Reding <thierry.reding@avionic-design.de>

This patch adds the -X option that used to be present in rdesktop and
which allows the xfreerdp window to be embedded within another X window.
---
 X11/xf_types.h |    1 +
 X11/xf_win.c   |    3 +++
 X11/xfreerdp.c |   14 ++++++++++++++
 3 files changed, 18 insertions(+), 0 deletions(-)

diff --git a/X11/xf_types.h b/X11/xf_types.h
index f40658f..450d555 100644
--- a/X11/xf_types.h
+++ b/X11/xf_types.h
@@ -53,6 +53,7 @@ struct xf_info
 	int keyboard_layout_id;
 
 	/* X11 stuff */
+	Window embed;
 	Window wnd;
 	GC gc;
 	Display * display;
diff --git a/X11/xf_win.c b/X11/xf_win.c
index 26563bb..f5c478f 100644
--- a/X11/xf_win.c
+++ b/X11/xf_win.c
@@ -1058,6 +1058,9 @@ xf_post_connect(xfInfo * xfi)
 		snprintf(win_title, sizeof(win_title), "%s:%d - freerdp", xfi->settings->server, xfi->settings->tcp_port_rdp);
 	XStoreName(xfi->display, xfi->wnd, win_title);
 
+	if (xfi->embed)
+		XReparentWindow(xfi->display, xfi->wnd, xfi->embed, 0, 0);
+
 	input_mask =
 		KeyPressMask | KeyReleaseMask | ButtonPressMask | ButtonReleaseMask |
 		VisibilityChangeMask | FocusChangeMask | StructureNotifyMask |
diff --git a/X11/xfreerdp.c b/X11/xfreerdp.c
index 1e7e0fd..fc59e8a 100644
--- a/X11/xfreerdp.c
+++ b/X11/xfreerdp.c
@@ -99,6 +99,7 @@ out_args(void)
 		"\t-f: fullscreen mode\n"
 		"\t-z: enable bulk compression\n"
 		"\t-x: performance flags (m, b or l for modem, broadband or lan)\n"
+		"\t-X: embed into another window with a given XID.\n"
 #ifndef DISABLE_TLS
 		"\t--no-tls: disable TLS encryption\n"
 #endif
@@ -337,6 +338,19 @@ process_params(xfInfo * xfi, int argc, char ** argv, int * pindex)
 			{
 				settings->performanceflags = strtol(argv[*pindex], 0, 16);
 			}
+		} else if (strcmp("-X", argv[*pindex]) == 0) {
+			*pindex = *pindex + 1;
+
+			if (*pindex == argc) {
+				printf("missing XID\n");
+				return 1;
+			}
+
+			xfi->embed = strtoul(argv[*pindex], NULL, 16);
+			if (!xfi->embed) {
+				printf("bad XID\n");
+				return 1;
+			}
 		}
 #ifndef DISABLE_TLS
 		else if (strcmp("--no-tls", argv[*pindex]) == 0)
